### DISCUSSION TITLE
Is it possible to inject previous history?

### DISCUSSION THREAD

**Discussion by tyler-suard-parker** [microsoft/autogen#2049] [Q&A]
Hello.  My frontend returns the last 4 exchanges with a chatbot.  I was wondering if there is a way to initialize an agent using that prior history.

**Comment by victordibia** [comment]
@tyler-suard-parker ,

Yes, you can update the history of agents. 

- loop through the history object 
- call `agent.send(message, receiver, request_reply=False)` . This adds the particular message to the agent history without triggering a reply. 
- finally, when you call initiate_chat, ensure you set `clear_history=False` ... `initiate_chat(clear_history=False)`


Some pseudo code below

```python

user_proxy = autogen.UserProxyAgent ..
assistant = autogen.AssistantAgent ...

for message in history:
    user_proxy.send(message, assistant, request_reply=False) 
    # you can also send from assistant to user_proxy depending on your use case

user_proxy.initiate_chat(assistant, message=message, clear_history=False)

```

**Comment by tyler-suard-parker** [comment]
@victordibia Dr. Dibia, this is SO helpful, thank you!

**Comment by easontsai** [comment]
how about groupchat? how to handle the chat history?

**Reply by jacinli** [reply]
I have a way to resolve history in GroupChat, but it will be a bit imperfect. You can write previous conversations into a history list.
For example:
`history_after_summary_str = " ".join([item["content"] for item in history]) GroupChat.DEFAULT_INTRO_MSG = history_after_summary_str`
Then you can use this writing method to initialize DEFAULT_INTRO_MSG of GroupChat, and set the send_introductions attribute in GroupChat to True.

**Reply by jacinli** [reply]
I now think of a better way, which is to not need to import the introduction into the group chat, but also to send my previous conversation records in advance. Here is my sample code, hope it helps.
` for msg in history:
                if msg["role"] == 2:
                 
                    for assistant in assistants:
                        await manager.a_send(msg["content"], recipient=assistant, request_reply=False, silent=True)
                # else:
                #     await manager.a_send(msg["content"], recipient=user_proxy, request_reply=False, silent=False)      
            #Send previous conversation information to manager before initchat 
            await user_proxy.a_initiate_chat(
                silent=True,
                recipient=manager,
                clear_history=False,
                message=message,
            )`

**Reply by ekzhu** [reply]
You can now resume a group chat from a previous session: https://microsoft.github.io/autogen/docs/topics/groupchat/resuming_groupchat

**Reply by jacinli** [reply]
Thanks!

**Comment by annusrcm** [comment]
> how about groupchat? how to handle the chat history?

Try this: 
https://clintgoodman27.medium.com/a-practical-guide-for-using-autogen-in-software-applications-8799185d27ee
**but with one change** :  await user_proxy.a_initiate_chat(manager, message=message, **clear_history=False)**
ensure clear_history is set to False
I have tried, it works for me

**Reply by ekzhu** [reply]
You can now resume a group chat from a previous session: https://microsoft.github.io/autogen/docs/topics/groupchat/resuming_groupchat

**Comment by afogarty85** [comment]
Anyone got an example of how we can include chat history clearly in a groupchat scenario?

**Reply by r4881t** [reply]
@afogarty85 - Something like this will work in a generic manner, providing an easy interface for user - agent/GroupChat/SoM chat history

https://github.com/microsoft/autogen/issues/2671#issue-2292415520

**Reply by ekzhu** [reply]
You can now resume a group chat from a previous session: https://microsoft.github.io/autogen/docs/topics/groupchat/resuming_groupchat
